File: src/main/java/com/woomulwoomul/woomulwoomulbackend/common/log/AspectLogging.java
package com.woomulwoomul.woomulwoomulbackend.common.log;

import com.woomulwoomul.woomulwoomulbackend.exception.CustomException;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.*;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;

@Slf4j
@Aspect
@Component
public class AspectLogging {
    @Pointcut("execution(public * com.woomulwoomul.woomulwoomulbackend.domain..*(..))" +
            "|| execution(public * com.gabojait.gabojaitspring.api..*(..))")
    private void global() {}

    @Pointcut("execution(protected * com.woomulwoomul.woomulwoomulbackend.exception.CustomExceptionHandler..*(..))")
    private void exception() {}

    @Before("global()")
    public void beforeGlobal(JoinPoint jp) {
        final MethodSignature signature = (MethodSignature) jp.getSignature();
        final String className = signature.getDeclaringType().getSimpleName();
        final Method method = signature.getMethod();
        final Parameter[] parameterNames = method.getParameters();
        final Object[] arguments = jp.getArgs();
        final int paramLength = Math.min(parameterNames.length, arguments.length);
        final String uuid = InterceptorLogging.getRequestId() == null ? "SYSTEM" : InterceptorLogging.getRequestId();

        StringBuilder param = new StringBuilder();
        for (int i = 0; i < paramLength; i++) {
            param.append(parameterNames[i].getName()).append("=");

            if (arguments[i] != null) param.append(arguments[i]);
            else param.append("null");

            if (i != parameterNames.length - 1) param.append(", ");
        }

        log.info("[{} | BEFORE] {} | {} ({})", uuid, className, method.getName(), param);
    }

    @AfterReturning(value = "global()", returning = "result")
    public void afterGlobal(JoinPoint jp, Object result) {
        final MethodSignature signature = (MethodSignature) jp.getSignature();
        final String className = signature.getDeclaringType().getSimpleName();
        final String methodName = signature.getMethod().getName();
        final String uuid = InterceptorLogging.getRequestId() == null ? "SYSTEM" : InterceptorLogging.getRequestId();

        log.info("[{} | AFTER] {} | {} | return={}", uuid, className, methodName, result);
    }

    @AfterThrowing(value = "exception()", throwing = "ex")
    public void afterThrowingGlobal(JoinPoint jp, CustomException ex) {
        final MethodSignature signature = (MethodSignature) jp.getSignature();
        final String className = signature.getDeclaringType().getSimpleName();
        final String methodName = signature.getMethod().getName();
        final String errorName = ex.getExceptionCode().name();
        final String uuid = InterceptorLogging.getRequestId() != null ? InterceptorLogging.getRequestId() : "SYSTEM";

        log.error("========== [{} | ERROR] {} | {} | code={} ==========", uuid, className, methodName, errorName);

        if (ex.getExceptionCode().getStatus().equals(HttpStatus.INTERNAL_SERVER_ERROR))
            if (ex.getThrowable() != null)
                log.error("========== [{} | DESCRIPTION] ==========", uuid, ex.getThrowable());
    }
}

File: core/src/main/kotlin/com/woomulwoomul/core/common/log/AspectLogging.kt
package com.woomulwoomul.core.common.log

import com.woomulwoomul.core.common.response.CustomException
import org.aspectj.lang.JoinPoint
import org.aspectj.lang.annotation.*
import org.aspectj.lang.reflect.MethodSignature
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR
import org.springframework.stereotype.Component

@Aspect
@Component
class AspectLogging {

    companion object {
        private val log = LoggerFactory.getLogger(this::class.java)
    }

    @Pointcut("execution(public * com.woomulwoomul.*.*..*(..))")
    private fun global() {}

    @Before("global()")
    fun beforeGlobal(jp: JoinPoint) {
        val signature = jp.signature as MethodSignature
        val className = signature.declaringTypeName.substringAfterLast(".")
        val method = signature.method
        val parameterNames = method.parameters
        val arguments = jp.args
        val paramLength = minOf(parameterNames.size, arguments.size)
        val uuid = InterceptorLogging().requestId

        val param = StringBuilder()
        for (i in 0 until paramLength) {
            param.append(parameterNames[i].name).append("=")
            param.append(arguments[i] ?: "null")

            if (i != parameterNames.size - 1) param.append(", ")
        }

        log.info("[{} | BEFORE] {} | {} ({})", uuid, className, method.name, param)
    }

    @AfterReturning(value = "global()", returning = "result")
    fun afterReturningGlobal(jp: JoinPoint, result: Any?) {
        val signature = jp.signature as MethodSignature
        val className = signature.declaringTypeName.substringAfterLast(".")
        val methodName = signature.method.name
        val uuid = InterceptorLogging().requestId
        var res = result.toString()

        if (methodName.contains("resultMasterPasswordScheduler"))
            res = res.replace(Regex("(?<=password\\s?=\\s?)\\S+"), "******")
                .replace(Regex("(?<=passwordReEnter\\s?=\\s?)\\S+"), "******")

        log.info("[{} | AFTER] {} | {} | return={}", uuid, className, methodName, res)
    }

    @AfterThrowing(value = "global()", throwing = "ex")
    fun afterThrowingGlobal(jp: JoinPoint, ex: CustomException) {
        val signature = jp.signature as MethodSignature
        val className = signature.declaringType.simpleName.substringAfterLast(".")
        val methodName = signature.method.name
        val errorName = ex.exceptionCode.name
        val uuid = InterceptorLogging().requestId

        log.error("========== [{} | ERROR] {} | {} | code={} ==========", uuid, className, methodName, errorName)

        if (ex.exceptionCode.httpStatus == INTERNAL_SERVER_ERROR)
            if (ex.throwable != null)
                log.error("========== [{} | DESCRIPTION] ==========", uuid, ex.throwable)
    }
}